Use request/response for data records

nfnt opened the pull request:
Data providers announce their available dataset using the DHT. Schedulers need a list of data slices of a dataset (= data record) if that dataset will be used in a job. We used to store this data record in the DHT as well. But DHT values are restricted in size and a data record can become quite large for some datasets. This commit changes the current behavior to no longer store data records in DHT. Instead, a data provider provides a request/response API for schedulers to request a data record for a dataset. By using request/response, we're no longer limited in the size a data record can have.

nfnt committed:
feat: use request/response for data records

Data providers announce their available dataset using the DHT.
Schedulers need a list of data slices of a dataset (= data record) if
that dataset will be used in a job. We used to store this data record in
the DHT as well. But DHT values are restricted in size and a data record
can become quite large for some datasets. This commit changes the
current behavior to no longer store data records in DHT. Instead, a data
provider provides a request/response API for schedulers to request a
data record for a dataset. By using request/response, we're no longer
limited in the size a data record can have.

orlandohohmeier commented on crates/data/src/bin/hypha-data.rs:
```diff
@@ -251,31 +251,49 @@ async fn run(config: ConfigWithMetadata<Config>) -> Result<()> {
 
     // Announce our dataset
     tracing::info!(dataset_name, "Announcing");
-    let _ = network.provide(dataset_name).await;
```
How do we discover dataset nodes?

nfnt replied:
The `network.provide` moved a few lines below and now has proper error handling. This change has been made to ensure we can handle data record request before we announce the dataset in the DHT.

Otherwise it's the same as before: By using `network.provide` the data providers peer id is added to the list of peers providing a specific dataset. Schedulers call `network.find_provider` to retrieve this list of peer ids.

orlandohohmeier replied:
got it, missed that line.

orlandohohmeier commented on crates/messages/src/lib.rs:
```diff
@@ -40,6 +41,7 @@ pub mod api {
         ParameterPull(parameter_pull::Response),
         ParameterPush(parameter_push::Response),
         Data(data::Response),
+        DataRecord(data_record::Response),
```
Let's please add a dedicated protocol (see actions, health) instead.

nfnt replied:
Why do we need a separate protocol? Isn't `hypha-api` meant for peer-to-peer communication?

orlandohohmeier replied:
Well, it's all peer-to-peer.... We started to separate the _API(s)_ so that we can individually version them and more importantly have a clearer connection between requests/response.

nfnt replied:
There's now a separate 'hypha-data-record' codec for scheduler-to-data-provider communication.

nfnt committed:
feat: use a separate codec for data records

Requesting data records from data providers is done using the new
'hypha-data-record' codec.

orlandohohmeier reviewed (approved):
LGTM
